New CSV implementation

Jolanrensen opened the pull request:
See the issue for the progress: https://github.com/Kotlin/dataframe/issues/827

~TODO:~
- [x] Split double parser into separate PR when FastDoubleParser publishes ConfigurableDoubleParser: https://github.com/Kotlin/dataframe/pull/935
- [x] Clean git history after merging fast double parser

Some high level overview of the changes:
- introduces new `:dataframe-csv` module
- new functions, requiring `@OptIn(ExperimentalCsv::class)`:
  - `read(Csv|Tsv|Delim)(Str)` based on Deephaven
  - `write(Csv|Tsv|Delim)`, and `to(Csv|Tsv|Delim)Str` based on Apache commons csv
- Added `ParserOptions.skipTypes`
- `parse` can now recognize `Char`
- Extended supported `ColTypes` to all the types we can parse
- Added OOM warning to old `readCSV` function informing the user about our new experimental functions :)
- See all checked off boxes in https://github.com/Kotlin/dataframe/issues/827 for all the bugs this fixes or issues it takes in mind.

Fixes https://github.com/Kotlin/dataframe/issues/787
Fixes https://github.com/Kotlin/dataframe/issues/508
Fixes https://github.com/Kotlin/dataframe/issues/589
Fixes https://github.com/Kotlin/dataframe/issues/469
Fixes https://github.com/Kotlin/dataframe/issues/921

Example of the new KDocs and all supported parameters:
![image](https://github.com/user-attachments/assets/d57c18db-c7c3-4f0a-a990-83e5fe7aec76)
![image](https://github.com/user-attachments/assets/5758da86-68e7-4ebe-819d-a21477378f1a)
![image](https://github.com/user-attachments/assets/191447f2-d248-4391-a661-e9e03af7ca4f)

Jolanrensen commented:
NOTE: maybe we can `try {} catch(OutOfMemoryError) {}` around the apache commons read csv and recommend the new experimental csv reader.

Jolanrensen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/api/parse.kt:
```diff
@@ -64,8 +67,17 @@ public data class ParserOptions(
     val dateTimeFormatter: DateTimeFormatter? = null,
     val dateTimePattern: String? = null,
     val nullStrings: Set<String>? = null,
+    val skipTypes: Set<KType> = emptySet(),
```
Maybe `skipTypes` be better be inverted to `checkTypes`. But that would mean exposing all `Parsers` so users/csv readers can filter which types to consider when parsing and which to skip. WDYT?

zaleslaw replied:
Really weird feature, let's live a little bit with that solution to reflect later in code

Jolanrensen replied:
It's mostly a performance thing, but the problem is that we don't have a user-facing way to refer to parsers, so we cannot say: "please parse this string column and try this type but not this type". I'm not sure how the API should look here

Jolanrensen commented:
Released dev version to test: `0.15.0-dev-4759`

Jolanrensen commented on dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/io/readDelim.kt:
```diff
@@ -0,0 +1,345 @@
+): DataFrame<*> {
+    // set up the csv specs
+    val csvSpecs = with(CsvSpecs.builder()) {
+        if (additionalCsvSpecs != null) from(additionalCsvSpecs)
```
Parameters set in `additionalCsvSpecs` are now overridden by the other (default) parameters. Do we think this is intended behavior? Or should we let `additionalCsvSpecs` overwrite all parameters given here?

Additionally, we could make a separate public function which takes _just_ a `CsvSpecs` instance and skips all our default parameters and DF logic. WDYT?

Jolanrensen replied:
Fixed by making it a lambda running on the builder :)

Jolanrensen commented on dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/io/ListSink.kt:
```diff
@@ -0,0 +1,202 @@
+ *
+ * The implementation is based on [Writing Your Own Data Sinks](https://github.com/deephaven/deephaven-csv/blob/main/ADVANCED.md).
+ *
+ * If we ever store column data unboxed / primitively, this needs to be modified.
```
Actually, we could also create a `ListSink` for each primitive type and `.asList()` the result. DataFrame can handle such columns. However, this could slow down processing after reading (as all values visited will need to be boxed) and potentially postpone the OOM exception. Unless the user filters first maybe...

Jolanrensen commented:
Benchmark results (in terms of speed; seconds per operation, lower is better):

| benchmark | type | score | scoreError | scoreUnit |
| :--- | :--- | :--- | :--- | :--- |
| apache | small | 0.005 | 0.001 | s/op |
| apache | medium | 3.397 | 0.224 | s/op |
| apache | large | OOM |  | s/op |
| deephaven | small | 0.009 | 0.002 | s/op |
| deephaven | medium | 1.278 | 0.109 | s/op |
| deephaven | large | 134.84 | 16.472 | s/op |

zaleslaw commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/annotations/ImportDataSchema.kt:
```diff
@@ -47,6 +47,7 @@ public enum class DataSchemaVisibility {
     EXPLICIT_PUBLIC,
 }
 
+// TODO add more options
```
TODO: convert to issues or fix

zaleslaw commented on dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/io/ListSink.kt:
```diff
@@ -0,0 +1,202 @@
+                }
+            }
+
+            else -> error("unsupported sink state")
```
Could you please improve the error message! Or give some advice

Jolanrensen replied:
The datatype supplied here can only be byte, short, int, or long, as described by https://github.com/deephaven/deephaven-csv/blob/main/ADVANCED.md#supporting-the-fast-path-for-numeric-type-inference, so this case never occurs, unless deephaven updates and does something weird. But I'll add a comment

zaleslaw commented on dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/io/readDelim.kt:
```diff
@@ -0,0 +1,343 @@
+ * Sets the correct parsers for the csv, based on [colTypes] and [ParserOptions.skipTypes].
+ * If [ColType.DEFAULT] is present, it sets the default parser.
+ */
+private fun CsvSpecs.Builder.parsers(
```
I also have one question, @Jolanrensen did you measure a JUnit test coverage here? When I am starting to think about multiple variety and combinatorial explosion of programming paths here, it makes me crazy.

Jolanrensen replied:
For this specific function, yes, all cases are covered.
If you narrow it down, there are fewer paths than you think:

- if no colTypes are given
  - let deephaven use all its default parsers
  - subtract parsers of skipTypes if those are supplied
- if colTypes are supplied
  - if DEFAULT is among the values
    - set the parser for each supplied column+coltype
    - let deephaven use _only_ the parser given as DEFAULT type
  - if DEFAULT is not among the values
    - set the parser for each supplied column+coltype
    - let deephaven use all its default parsers
    - subtract parsers of skipTypes if those are supplied

(you may notice that `skipTypes` will never skip a type explicitly set by `colTypes`. This is intended.)

I will add this overview to the kdoc too, It may be helpful when debugging.

I will add a failing date test though. (if it fails, it remains String)

zaleslaw commented on dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/Compression.kt:
```diff
@@ -0,0 +1,67 @@
+ *
+ * @param wrapStream function that wraps any [InputStream] into a decompressing [InflaterInputStream] stream
+ */
+public sealed class Compression<I : InputStream>(public open val wrapStream: (InputStream) -> I) {
```
I think our core users will be happy with compression support

Jolanrensen replied:
Will move it after dataframe-csv becomes non-experimental

zaleslaw commented on dataframe-csv/src/test/kotlin/org/jetbrains/kotlinx/dataframe/io/BenchmarkTest.kt:
```diff
@@ -0,0 +1,54 @@
+
+    @Setup
+    fun setup() {
+        System.setProperty("org.slf4j.simpleLogger.defaultLogLevel", "info")
```
it's really cool!

zaleslaw commented on dataframe-csv/src/test/kotlin/org/jetbrains/kotlinx/dataframe/io/DelimCsvTsvTests.kt:
```diff
@@ -0,0 +1,657 @@
+    }
+
+    @Test
+    fun `read 2 compressed Csv`() {
```
Do you have a test for reading something zipped, but that is not CSV or has an extension, but in reality has a weird format internally

zaleslaw commented on dataframe-csv/src/test/resources/largeCsv.csv.gz:
I'm not sure that we should have this file in our repo just for performance test, this is a reason why I've excluded SQL dump from the repo

Let's discuss what it's possible to do, but it's fine
1) write a csv generator programm, which creates a file in tmp directory
2) save it externally and add a link for loading

Jolanrensen replied:
generating it sounds interesting. 
Storing it externally, while I'm normally against doing that for tests, might work for this case, since benchmarks are not run as part of the normal test suite.

Jolanrensen replied:
that said, I can also just remove the benchmarking test altogether. What do you think?

zaleslaw replied:
Let's remove!

koperagen commented on dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/io/DataFrameCustomDoubleParser.kt:
```diff
@@ -0,0 +1,24 @@
+
+    override fun parse(cs: CharSequence): Double =
+        fastDoubleParser.parseOrNull(cs.toString())
+            ?: throw NumberFormatException("Failed to parse double")
```
maybe it will be useful to include substring to exception message

Jolanrensen replied:
the message is already debug-logged in parseOrNull.
This `NumberFormatException` is just for Deephaven to catch. Actually, I can even remove the message altogether, as Deephaven doesn't do anything with it either. (they just catch it and return false)

koperagen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/common.kt:
```diff
@@ -1,28 +1,46 @@
+ *
+ * Public so it may be used in other modules.
+ */
+public fun catchHttpResponse(url: URL, body: (InputStream) -> AnyFrame): AnyFrame {
```
Did you consider other ways to expose this API. Is it somehow possible to keep in a separate artifact so it's only available with explicit dependency?

Jolanrensen replied:
friend modules would work, but I'm not entirely a fan of those. A separate module could also work maybe. I could think of some more functions/classes that would benefit from being in a separate internal module (code generation, the double parser, these file/url/path/string helper functions, etc.) But that might be better solved in another issue.

koperagen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/csv.kt:
```diff
@@ -331,58 +355,72 @@ public fun DataFrame.Companion.readDelim(
-        var count = readLines ?: 0
-        while (iter.hasNext() && 0 < count--) {
-            records.add(iter.next())
+    try {
```
Please extract try body in a readDelimImpl

koperagen commented on core/src/test/kotlin/org/jetbrains/kotlinx/dataframe/io/CsvTests.kt:
```diff
@@ -70,7 +70,7 @@ class CsvTests {
         df.columnNames()[5] shouldBe "duplicate1"
         df.columnNames()[6] shouldBe "duplicate11"
-        df["duplicate1"].type() shouldBe typeOf<String?>()
+        df["duplicate1"].type() shouldBe typeOf<Char?>()
```
So it converts columns with single characters to Char?

Jolanrensen replied:
Deephaven has the same functionality, so I thought it would be a good idea to make our parsers behave the same

koperagen commented on dataframe-csv/build.gradle.kts:
```diff
@@ -0,0 +1,170 @@
+kotlin {
+    explicitApi()
+    sourceSets.all {
+        languageSettings {
```
leftovers

koperagen commented on dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/documentation/DelimParams.kt:
```diff
@@ -0,0 +1,245 @@
+     *   If `true`, the data will be parsed in parallel.
+     *   This is usually faster, but can be turned off for debugging.
+     */
+    const val PARSE_PARALLEL: Boolean = true
```
i feel it's a big ambiguous. So it's not about deephaven parser, rather it's about DataColumn.parse?

Jolanrensen replied:
no it's deephaven, our parser is not parallel atm. I'll specify.

koperagen commented on dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/Compression.kt:
```diff
@@ -0,0 +1,105 @@
+ *
+ * @param wrapStream function that wraps any [InputStream] into a decompressing [InflaterInputStream] stream
+ */
+public sealed class Compression<I : InputStream>(public open val wrapStream: (InputStream) -> I) {
```
Is this type parameter needed? In the end readCsv has Compression<*>, so it's the same as `Compression(public open val wrapStream: (InputStream) -> InputStream)`

koperagen commented on dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/Compression.kt:
```diff
@@ -0,0 +1,105 @@
+    }
+}
+
+public fun compressionStateOf(fileOrUrl: String): Compression<*> =
```
Not sure about "state". 
I'd expect this factory to be in companion object of Compression, not on top level

koperagen commented on dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/readCsv.kt:
```diff
@@ -0,0 +1,292 @@
+    path: Path,
+    delimiter: Char = CSV_DELIMITER,
+    header: List<String> = HEADER,
+    hasFixedWidthColumns: Boolean = HAS_FIXED_WIDTH_COLUMNS,
```
Are hasFixedWidthColumns and fixedColumnWidths connected? What will happen if i provide false and non-empty list?

Jolanrensen replied:
Then `fixedColumnWidths` will be ignored. They are only set if `hasFixedWidthColumns == true`.

Deephaven throws an error in this case. We could also do that if you like

koperagen commented on dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/io/Compression.kt:
```diff
@@ -0,0 +1,105 @@
+    /**
+     * No compression.
+     */
+    public data object None : Compression<InputStream>(wrapStream = { it })
```
Seems like "None" is actually kind of "Detect", no? At least according to kdocs it seems like it is

Jolanrensen replied:
no, it's actually "none". If a user supplies `compression = Compression.None`, the inputStream will be supplied to deephaven-csv directly.

If a user does _not_ supply a compression argument, the default will be "`Compression.None`, unless detected otherwise from the input file or url." So `compressionStateOf()` will be called (if possible) and the right one will be chosen based on the file extension, defaulting to `None` if that fails.

Hope that clarifies it :) If not, please tell me which part can be written clearer

koperagen commented on dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/io/readDelim.kt:
```diff
@@ -0,0 +1,343 @@
+            if (e.message ==
+                "Can't proceed because hasHeaderRow is set but input file is empty or shorter than skipHeaderRows"
+            ) {
+                return@readDelimImpl DataFrame.empty()
```
Are you sure about it? I incline to think that this is an error that needs to be fixed, having and empty dataframe instead of the exception will be not-so-clear

Jolanrensen replied:
It was to align with the old behavior: feeding an empty file/string should result in an empty dataframe (we could change this of course).

Deephaven can do this successfully if you supply a header yourself (it simply returns empty columns), however it throws an exception when it needs to guess a header from an empty file, like here. But if the file is empty anyway, it makes no sense to require the user to supply a header to avoid this exception, so that's why I caught it like this.

I also tried detecting whether the inputstream is empty before passing it to CsvReader, but I didn't find a reliable way (`.available()` can return `0`, even though the stream still has data).

Jolanrensen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/api/parse.kt:
```diff
@@ -388,6 +388,8 @@ internal object Parsers : GlobalParserOptions {
             }
         },
+        // Char
+        stringParser<Char> { it.singleOrNull() },
```
not necessarily, columns with single characters can be quite common, especially when reading pivoted data, plus it could save memory, which is always a good thing, right? :)

koperagen commented on dataframe-csv/src/test/resources/abc.tsv:
```diff
@@ -0,0 +1,3 @@
+A	B	C
```
Maybe it'd be ok to just read files from `../core/src/test/resources`?

Jolanrensen replied:
maybe yeah, but I think once we remove the old csv implementation from :core we can better remove the tests and resources there. Then they're not duplicate anymore

koperagen commented on dataframe-csv/src/test/kotlin/org/jetbrains/kotlinx/dataframe/io/DelimCsvTsvTests.kt:
```diff
@@ -0,0 +1,657 @@
+            val df = DataFrame.readCsv(wineCsv, delimiter = ';')
+            val schema = df.schema()
+
+            fun assertColumnType(columnName: String, kClass: KClass<*>) {
```
duplicated function

koperagen commented on dataframe-csv/src/test/kotlin/org/jetbrains/kotlinx/dataframe/io/DelimCsvTsvTests.kt:
```diff
@@ -0,0 +1,657 @@
+    fun `check integrity of example data`() {
+        val df = DataFrame.readCsv(
+            "../data/jetbrains_repositories.csv",
+            skipLines = 1, // now needs this, ignoreEmptyLines cannot catch it
```
let's update the data?

Jolanrensen replied:
I think it's good that we know the difference in behavior for the new implementation; what happens if we try to read a csv file with a blank line at the top.

Jolanrensen commented on plugins/symbol-processor/src/main/kotlin/org/jetbrains/dataframe/ksp/DataSchemaGenerator.kt:
```diff
@@ -155,6 +155,7 @@ class DataSchemaGenerator(
             codeGenerator.createNewFile(Dependencies(true, importStatement.origin), packageName, "$name.Generated")
 
         val formats = listOf(
+            // TODO new Csv() and Tsv()
```
It's already part of the umbrella issue todo-list: https://github.com/Kotlin/dataframe/issues/827

koperagen commented on dataframe-csv/src/test/kotlin/org/jetbrains/kotlinx/dataframe/io/DelimCsvTsvTests.kt:
```diff
@@ -0,0 +1,657 @@
+
+    @Test
+    fun `read GZ Csv`() {
+        val df = DataFrame.readCsv(simpleCsvGz)
```
I think a lot of these assertions aren't needed. We only need to test that reading .gz works the same as reading plain csv

Jolanrensen replied:
yeah that's fair

koperagen commented on dataframe-csv/src/test/kotlin/org/jetbrains/kotlinx/dataframe/io/DelimCsvTsvTests.kt:
```diff
@@ -0,0 +1,657 @@
+    }
+
+    @Test
+    fun `if string starts with a number, it should be parsed as a string anyway`() {
```
I wanted to see actual file to understand what name implies. Maybe it'd be better to inline the file and use readCsvStr

koperagen commented on dataframe-csv/src/test/kotlin/org/jetbrains/kotlinx/dataframe/io/DelimCsvTsvTests.kt:
```diff
@@ -0,0 +1,657 @@
+import kotlin.reflect.typeOf
+
+@Suppress("ktlint:standard:argument-list-wrapping")
+class DelimCsvTsvTests {
```
We already have a lot of CSV tests in core tests. Do you copy them here or this is a completely written from scratch? I understand that API is different, but at least for default case when only path argument is supplied, it'd be good to compare results and to be able to say they're identical or not

Jolanrensen replied:
yes I copied all of them and added a couple more, so I could see they behave the same.

Jolanrensen replied:
and actually, I changed the default arguments such that the behavior of not specifying any additional settings is as close as possible as the old behavior. Things like `allowMissingColumns` ans `ignoreExcessColumns` for instance.

Jolanrensen commented on dataframe-csv/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/io/readDelim.kt:
```diff
@@ -0,0 +1,343 @@
+
+        // Deephaven's LocalDateTime parser is unconfigurable, so if the user provides a locale, pattern, or formatter,
+        // we must use our own parser for LocalDateTime and let Deephaven read them as Strings.
+        val useDeepHavenLocalDateTime = with(parserOptions) {
```
Probably yes, however, in theory the user should not be able to tell the difference, only in speed.

- If they provide no configuration, it will use the fast LocalDateTime parser of deephaven and if that fails, it will remain a String and our slower LocalDateTime parser will try it too.

- If they do provide configuration, then deephaven's parser cannot be used and our slower LocalDateTime parser will be used.

In both cases, the user will get the expected result and where possible we save some performance. But we cannot "recommend" to not specify a configuration to gain performance, because that's dependent on the data the user has to read. All we can do is mention it somewhere in the documentation.

Jolanrensen commented:
@zaleslaw @koperagen My two latest commits (https://github.com/Kotlin/dataframe/compare/703338a7042c30f8629bfc9e5979c0a6560a067f..53e2f64daa1b8167c56d5c4a4fd33a66cf78bfe2) contain a lot of new changes because I figured out how `ParserOptions` actually work in the library:
- There's a global setting `DataFrame.parser` which should be respected if either: `ParserOptions == null` or any of its parameters are `null`. I made sure the csv readers respect this too.
- I added my new `ParserOptions` arguments to the global one as well (and I made it not a data class anymore for better future binary compatibility).
- "skip types except" makes no sense, we have `convertTo` for that. But there was no overload for `stringColumn.convertTo` that took any `ParserOptions`, so I added that.
- Moved `convertToDouble` to `convertToDoubleImpl` and added explanation in KDocs for why (I think) it exists (and it now accepts `nullStrings` as well).
- Added more relevant tests

Could you rereview if you like?

koperagen commented on core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/api/parse.kt:
```diff
@@ -293,14 +306,19 @@ internal object Parsers : GlobalParserOptions {
+    // same as parserToDoubleWithOptions, but overrides the locale to C.UTF-8
+    private val posixParserToDoubleWithOptions = stringParserWithOptions { options ->
+        val parserOptions = (options ?: ParserOptions()).copy(locale = Locale.forLanguageTag("C.UTF-8"))
+        TODO()
```
something is not right here

Jolanrensen replied:
oops thanks

zaleslaw commented:
Is it a first global setting in KDF, or we had something earlier? Should user set up it manually and if it will be replaced is it replaced for the whole project, using our library? In that way it's closer to the CONFIG file which is visible or Spring Bean.

I'm ok, but looks like a temporary solution, could you create a ticket with research tag for that

Jolanrensen commented:
> Is it a first global setting in KDF, or we had something earlier? Should user set up it manually and if it will be replaced is it global for the whole project, using our library? In that way it's closer to the CONFIG file which is visible or Spring Bean.
> 
> I see that it's inherited from external component, isn't it?
> 
> I'm ok, but looks like a temporary solution, could you create a ticket with research tag for that

`DataFrame.parser` has been part of the library for years, I just didn't know about it yet (it's also part of the documentation https://kotlin.github.io/dataframe/parse.html, at the bottom). It's similar to `Locale.setDefault()` on the JVM, so yes, it's global, like a config file, but it can be changed programmatically because it's stored in a singleton object.

In all cases, the global default can also be overridden by providing a `ParserOptions` instance to the function, so we embrace both a functional and imperative style which I think is cool.
